pkg/rm/tcc/fence/handler/tcc_fence_dbproxy_handler.go

```
@@ -0,0 +1,213 @@
+	xid := tm.GetBusinessActionContext(ctx).Xid
+	branchId := tm.GetBusinessActionContext(ctx).BranchId
+	actionName := tm.GetBusinessActionContext(ctx).ActionName
+
+	defer func() {
+		rec := recover()
+		err, ok := rec.(seataErrors.TccFenceError)
+		if ok {
+			if err.Code == seataErrors.FenceErrorCodeDuplicateKey {
+				handler.AddToLogCleanQueue(xid, branchId)
+			}
+		}
+		// panic uniform process in outside method
+		panic(rec)
+	}()
+
+	ok := handler.InsertTCCFenceLog(tx, xid, branchId, actionName, constant.StatusTried)
+	log.Infof("tcc fence prepare result: %b. xid: %s, branchId: %d", ok, xid, branchId)
+	if ok {
+		err := callback()
+		if err != nil {
+			if rollbackErr := tx.Rollback(); rollbackErr != nil {
+				return errors.New(err.Error() + rollbackErr.Error())
+			}
+			return err
+		}
+		return tx.Commit()
+	} else {
+		return seataErrors.NewTccFenceError(seataErrors.InsertRecordError, fmt.Sprintf("insert tcc fence record errors, prepare fence failed. xid= %s, branchId= %d", xid, branchId))
+	}
+}
+
+func (handler *tccFenceDbProxyHandler) CommitFence(ctx context.Context, tx *sql.Tx, callback func() error) error {
+	xid := tm.GetBusinessActionContext(ctx).Xid
+	branchId := tm.GetBusinessActionContext(ctx).BranchId
+
+	fenceDo := handler.tccFenceDao.QueryTCCFenceDO(tx, xid, branchId)
+	if fenceDo == nil {
+		return seataErrors.NewTccFenceError(seataErrors.RecordNotExists, fmt.Sprintf("tcc fence record not exists, commit fence method failed. xid= %s, branchId= %d ", xid, branchId))
+	}
+	if fenceDo.Status == constant.StatusCommitted {
+		log.Infof("branch transaction has already committed before. idempotency rejected. xid: %s, branchId: %d, status: %d", xid, branchId, fenceDo.Status)
+		return nil
+	}
+	if fenceDo.Status == constant.StatusRollbacked || fenceDo.Status == constant.StatusSuspended {
+		// enable warn level
+		log.Warnf("Branch transaction status is unexpected. xid: %s, branchId: %d, status: %s", xid, branchId, fenceDo.Status)
```

I recommend to state what you are currently doing when something unexpected happens

---

> I don't know much about golang，there are something have to be aware of.
> 
> 1. All pre、commit、rollback need to be completed in one transaction.
> 2. Each resource could has a unique configuration of fence.
>    If we have done these things, please ignore.

这两件事都做了

---

The logs here are consistent with the seata-java implementation

---

It might be better to return an error here, so I'll give it a try